teps/0090-matrix.md

```diff
@@ -596,17 +627,546 @@ For example:
+
+### Results from Matrixed PipelineTasks
+
+Today, each `PipelineTask` is executed in a single `TaskRun` or `Run` which produces 
+the `Results` specified in the underlying `Task` or `Custom Task`. The `Results` are 
+accessible through variable - `$(tasks.<pipelinetask-name>.results.<result-name>)`. 
+
+With the `Matrix`, the `PipelineTask` will be executed in several `TaskRuns` or `Runs`,
+each of which will produce its own `Results`. Therefore, we need to adjust the variables
+for `Results` when a `Matrix` is present to accommodate all the `Results` produced. 
+We propose suffixing the `PipelineTask` name in the `Result` variable with the count 
+of the matrix execution: `$(tasks.<pipelinetask-name>-<count>.results.<result-name>)`.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline
+spec:
+  params:
+  - name: browsers
+    type: array
+    default:
+      - chrome
+      - firefox
+      - safari
+  tasks:
+  - name: clone
+    taskRef:
+      name: git-clone
+    ...
+  - name: test
+    taskRef:
+      name: test
+    matrix:
+    - name: browsers
+      value: $(params.browsers)
+    ...
+  - name: notify
+    taskRef:
+      name: notify
+    params:
+    - name: messages
+      value: 
+      - $(tasks.test-1.results.output)
+      - $(tasks.test-2.results.output)
+      - $(tasks.test-3.results.output)
+    ...
+```
+
+However, this presents a challenge when the matrix is dynamic (uses `Results`) because
+the count cannot be predetermined, making it hard to reference those `Results` in
+subsequent `PipelineTasks` at authoring time.
+
+In [TEP-0076: Array Results and Indexing][tep-0076], we are exploring supporting `Results`
+of type `Array` with indexing. That would allow us to produce `Results` of type `Array` 
+from the matrixed `PipelineTasks`, where the name of the array `Result` would be the 
+same as is today - `$(tasks.<pipelinetask-name>.results.<result-name>)`. 
+The array `Result` would enable iterating through all the `Results` from the fan out, 
+and indexing would enable choosing `Results` from specific `TaskRuns` or `Runs`.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline
+spec:
+  params:
+  - name: browsers
+    type: array
+    default:
+  tasks:
+  - name: clone
+    taskRef:
+      name: git-clone
+    ...
+  - name: get-test-shards
+    taskRef:
+      name: get-test-shards
+    runAfter:
+    - clone
+    ...
+  - name: test
+    taskRef:
+      name: test
+    matrix:
+    - name: shards
+      value: $(tasks.get-test-shards.results.shards)
+    ...
+  - name: notify
+    taskRef:
+      name: notify
+    params:
+    - name: messages
+      value: $(tasks.test.results.output)
+    ...
+```
+
+After [TEP-0076][tep-0076] is implemented, we will explore this further in
+another TEP. In the meantime, we will disallow consuming `Results` from a
+dynamic matrixed `PipelineTask` given the limitation discussed above.
```

Comment:
This limitation is really limiting 😓 
I agree that there is no easy way around it, without having more structured results, but we could consider workarounds, like producing a serialised JSON representation of the results. It gets complicated easily with more dimensions though...

Reply:
It seems that we are making great progress in [TEP-0075](https://github.com/tektoncd/community/pull/479) and [TEP-0076](https://github.com/tektoncd/community/pull/477) so I'd prefer that we wait for the structured Results proposed in those TEPs to be available instead of providing a temporary workaround as we wait for them (cc @bobcatfish)